[SPARK-8603][SPARKR] Use shell() instead of system2() for SparkR on Windows

HyukjinKwon opened the pull request:
## What changes were proposed in this pull request?

This PR corrects SparkR to use `shell()` instead of `system2()` on Windows.

Using `system2(...)` on Windows does not process windows file separator `\`. `shell(tralsate = TRUE, ...)` can treat this problem. So, this was changed to be chosen according to OS.

Existing tests were failed on Windows due to this problem. For example, those were failed.

```
8. Failure: sparkJars tag in SparkContext (@test_includeJAR.R#34)
9. Failure: sparkJars tag in SparkContext (@test_includeJAR.R#36)
```

The cases above were due to using of `system2`.

In addition, this PR also fixes some tests failed on Windows.

```
5. Failure: sparkJars sparkPackages as comma-separated strings (@test_context.R#128) 
6. Failure: sparkJars sparkPackages as comma-separated strings (@test_context.R#131) 
7. Failure: sparkJars sparkPackages as comma-separated strings (@test_context.R#134) 
```

  The cases above were due to a weird behaviour of `normalizePath()`. On Linux, if the path does not exist, it just prints out the input but it prints out including the current path on Windows.

``` r
# On Linus
path <- normalizePath("aa")
print(path)
[1] "aa"

# On Windows
path <- normalizePath("aa")
print(path)
[1] "C:\\Users\\aa"
```
## How was this patch tested?

Jenkins tests and manually tested in a Window machine as below:

Here is the [stdout](https://gist.github.com/HyukjinKwon/4bf35184f3a30f3bce987a58ec2bbbab) of testing.

Closes #7025

pcp-dev committed:
[SPARK-8603] [sparkR] In windows, Incorrect file seperator passed to Java and Scripts from R

HyukjinKwon committed:
Fetch upstream

HyukjinKwon committed:
Correct condition, fix style and comments

HyukjinKwon commented:
Please let me cc @sun-rui and @JoshRosen

SparkQA commented:
**[Test build #58758 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58758/consoleFull)** for PR 13165 at commit [`454b7f5`](https://github.com/apache/spark/commit/454b7f53c1bf587325154535204695247c1b3d6a).

HyukjinKwon committed:
Fetch upstream

sun-rui commented:
Could you add a test case for Windows specific test? @shivaram

HyukjinKwon committed:
Update indentation

SparkQA commented:
**[Test build #58760 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58760/consoleFull)** for PR 13165 at commit [`3057844`](https://github.com/apache/spark/commit/3057844fbf406cf492638b8078e642139a199d51).

SparkQA commented:
**[Test build #58758 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58758/consoleFull)** for PR 13165 at commit [`454b7f5`](https://github.com/apache/spark/commit/454b7f53c1bf587325154535204695247c1b3d6a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58758/
Test PASSed.

SparkQA commented:
**[Test build #58760 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58760/consoleFull)** for PR 13165 at commit [`3057844`](https://github.com/apache/spark/commit/3057844fbf406cf492638b8078e642139a199d51).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58760/
Test PASSed.

srowen commented on R/pkg/R/client.R:
```diff
@@ -43,6 +43,17 @@ determineSparkSubmitBin <- function() {
+# of the operating system. So, this should be manaully changed.
+determinefileSeperator <- function() {
+  if (.Platform$OS.type == "windows") {
+    fileSeperator <- "\\"
```
Nit: separator not seperator

HyukjinKwon replied:
Thank you so much.

HyukjinKwon committed:
Fix typo

SparkQA commented:
**[Test build #58763 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58763/consoleFull)** for PR 13165 at commit [`935d796`](https://github.com/apache/spark/commit/935d796039a74e068b954b4c331a2b754bbdbede).

SparkQA commented:
**[Test build #58763 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58763/consoleFull)** for PR 13165 at commit [`935d796`](https://github.com/apache/spark/commit/935d796039a74e068b954b4c331a2b754bbdbede).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58763/
Test PASSed.

felixcheung commented:
Does this apply to other cases:
https://github.com/apache/spark/blob/d6dc12ef0146ae409834c78737c116050961f350/R/pkg/inst/worker/daemon.R#L22
https://github.com/apache/spark/blob/d6dc12ef0146ae409834c78737c116050961f350/R/pkg/inst/profile/shell.R#L20
https://github.com/apache/spark/blob/6ab4d9e0c76b69b4d6d5f39037a77bdfb042be19/examples/src/main/r/dataframe.R#L37

(last one is an example)

HyukjinKwon commented:
@sun-rui @felixcheung Let me try to build and run all tests for R first in Windows and then will try to correct and add each test one by one. This will take a bit of time and I might have to ask a lot of questions but anyway I will try.

sun-rui commented:
@felixcheung, this issue seems to relate to system2() only. However, let's wait for HyukjinKwon's test result.

@HyukjinKwon, great, go ahead please.

HyukjinKwon commented:
@sun-rui @felixcheung Right. It seems finally I made it. I made gists and upload a PDF file for Spark UI.

Let me tell you the test results first.

Here is the stdout output for the tests on Windwos 7 32bit, [output.msg](https://gist.github.com/HyukjinKwon/6a10719d2ca67e04ece2b23a8f92dc62).

Here is the stderr output for the tests on Windwos 7 32bit, [output.err](https://gist.github.com/HyukjinKwon/54984d57ee18236d46e965d07b31f77a).

Here is the PDF for [Spark UI PDF](https://drive.google.com/open?id=0B7RfLjRU7QTnVVA2bkVMVFkzNEE)

1. I run tests after building Spark on Windows according to [`./R/WINDOWS.md`] (https://github.com/apache/spark/blob/master/R/WINDOWS.md)

2. It seems `$HADOOP_HOME` should be set.

3. It seems `winutils.exe` is required which is included in Hadoop official binary although it reads file in the local file system.

4. And run the tests by the command below:

```bash
cd bin
spark-submit2.cmd --conf spark.hadoop.fs.defualt.name="file:///" ..\R\pkg\tests\run-all.R > output.msg 2> output.err
```

HyukjinKwon commented:
This raises some questions to me.

1. It seems several tests were failed. Could you please inform me your thoughts?

2. Now, I think I can add some tests but could you please where I should write the related tests and maybe rough ideas of the tests I should add?

sun-rui commented:
1. A rough scan of the test failures shows most of them are probably related to path handling. You can replay the failed test case in R on Windows.  For debug facilities in R, refer to http://www.inside-r.org/r-doc/base/traceback, https://stat.ethz.ch/R-manual/R-devel/library/base/html/debug.html, http://www.inside-r.org/r-doc/base/browser
2. You can add a new file named test_Windows.R under R/pkg/inst/tests/testthat
3. Sure.

sun-rui commented:
I checked the source code of system2. Since a batch file with .cmd extension is to be launched on Windows, I think it is more formal to use shell() instead of system2(). Refer to http://www.astrostatistics.psu.edu/datasets/R/html/base/html/shell.html. The proposed fix is like follows:
```
if (.Platform$OS.type == "windows") {
  shell(sparkSubmitBin, translate = TRUE, ...)
} else {
  system2(sparkSubmitBin, ...)
}
```

HyukjinKwon commented:
@sun-rui Thank you so much. I will try to add a test.

shivaram commented:
Related to my comment on #13217 -- I will test this out on windows using the new instructions.

HyukjinKwon commented:
@shivaram Thank you so much. Let me try to add a test meanwhile.

HyukjinKwon committed:
Address comments, fix failing tests on windos, add a test.

HyukjinKwon committed:
Skip Windows test on Linux and repleace jars to list

HyukjinKwon commented:
@shivaram and @sun-rui I addressed the comment, added a test for this PR and fixed some tests not working on Windows due to path separator.

This PR also fixes the test failures below.

- **Fixed case 1**

  ```
5. Failure: sparkJars sparkPackages as comma-separated strings (@test_context.R#128) 
6. Failure: sparkJars sparkPackages as comma-separated strings (@test_context.R#131) 
7. Failure: sparkJars sparkPackages as comma-separated strings (@test_context.R#134) 
```

  The cases above were due to a weird behaviour of `normalizePath()`. On Linux, if the path does not exist, it just prints out the input but it prints out including the current path on Windows.

  ```r
# On Linus
path <- normalizePath("aa")
print(path)
[1] "aa"

# On Windows
path <- normalizePath("aa")
print(path)
[1] "C:\\Users\\aa"
```

- **Fixed case 2**

  ```
8. Failure: sparkJars tag in SparkContext (@test_includeJAR.R#34)
9. Failure: sparkJars tag in SparkContext (@test_includeJAR.R#36)
```

The cases above were due to using of `system2`.

Here is full stdout of my testing, [output.msg](https://gist.github.com/HyukjinKwon/c72b747a15948bd062a26aba7c42115b) on Windows.

SparkQA commented:
**[Test build #59275 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59275/consoleFull)** for PR 13165 at commit [`64141d2`](https://github.com/apache/spark/commit/64141d20049098c51ac7dd99c38b840b25ca073a).

SparkQA commented:
**[Test build #59275 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59275/consoleFull)** for PR 13165 at commit [`64141d2`](https://github.com/apache/spark/commit/64141d20049098c51ac7dd99c38b840b25ca073a).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59275/
Test FAILed.

HyukjinKwon committed:
Fix R style

SparkQA commented:
**[Test build #59276 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59276/consoleFull)** for PR 13165 at commit [`4a53360`](https://github.com/apache/spark/commit/4a53360b6917a4851f8a2dff61f6e706ea6470f9).

SparkQA commented:
**[Test build #59276 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59276/consoleFull)** for PR 13165 at commit [`4a53360`](https://github.com/apache/spark/commit/4a53360b6917a4851f8a2dff61f6e706ea6470f9).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59276/
Test FAILed.

sun-rui commented on R/pkg/R/client.R:
```diff
@@ -60,6 +60,15 @@ generateSparkSubmitArgs <- function(args, sparkHome, jars, sparkSubmitOpts, pack
   combinedArgs
 }
 
+determineLauncher <- function(sparkSubmitBin, combinedArgs, capture = FALSE) {
```
can this be a common utility function and be re-used in test_includeJAR.R?

HyukjinKwon replied:
Sure. I will. 

BTW, could you inform me if I can disable style checking for some blocks? It seems it fails due to the usages of `shell`.

shivaram replied:
You can use `nolint` - See `context.R` for an example. BTW what is the lint error here ?

HyukjinKwon replied:
@shivaram For some reasons, it seems there is no mothod `shell()` on Linux (I tested this on my Mac) but on Windows. Here is a thread for this, http://r.789695.n4.nabble.com/Shell-Function-not-on-Linux-td4672569.html.

I could not find the official documentation though. Let me address the comments first.

sun-rui replied:
the common function could be like:
```
  launchScript(script) {
    if (windowsOS) use shell()
    else (system2()
  }
```
This makes it more general, not binding to launching SparkSubmit

HyukjinKwon replied:
@sun-rui Thank you so much. I was trying to come up with a good name.

HyukjinKwon committed:
Fix logical bool for capture (to make sure it waits)

SparkQA commented:
**[Test build #59277 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59277/consoleFull)** for PR 13165 at commit [`397fd14`](https://github.com/apache/spark/commit/397fd1429e52a6f856c32851fc4c8696028867c1).

SparkQA commented:
**[Test build #59277 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59277/consoleFull)** for PR 13165 at commit [`397fd14`](https://github.com/apache/spark/commit/397fd1429e52a6f856c32851fc4c8696028867c1).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59277/
Test FAILed.

HyukjinKwon committed:
Address comments

HyukjinKwon committed:
R style improvement

HyukjinKwon commented:
I just addressed the comments and here is the (stdout)[https://gist.github.com/HyukjinKwon/4bf35184f3a30f3bce987a58ec2bbbab] of testing on Windiws.

SparkQA commented:
**[Test build #59336 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59336/consoleFull)** for PR 13165 at commit [`0482ebb`](https://github.com/apache/spark/commit/0482ebbc43ff1bef8e7a6a16376c6ec36840a366).

SparkQA commented:
**[Test build #59336 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59336/consoleFull)** for PR 13165 at commit [`0482ebb`](https://github.com/apache/spark/commit/0482ebbc43ff1bef8e7a6a16376c6ec36840a366).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59336/
Test FAILed.

HyukjinKwon commented:
retest this please

SparkQA commented:
**[Test build #59339 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59339/consoleFull)** for PR 13165 at commit [`0482ebb`](https://github.com/apache/spark/commit/0482ebbc43ff1bef8e7a6a16376c6ec36840a366).

SparkQA commented:
**[Test build #59339 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59339/consoleFull)** for PR 13165 at commit [`0482ebb`](https://github.com/apache/spark/commit/0482ebbc43ff1bef8e7a6a16376c6ec36840a366).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59339/
Test PASSed.

sun-rui commented on R/pkg/inst/tests/testthat/test_includeJAR.R:
```diff
@@ -21,10 +21,13 @@ runScript <- function() {
-  res <- system2(command = submitPath,
-                 args = c(jarPath, scriptPath),
-                 stdout = TRUE)
+  if (.Platform$OS.type == "windows") {
```
you can call determineSparkSubmitBin() here

HyukjinKwon committed:
Reuse determineSparkSubmitBin()

HyukjinKwon committed:
Fix style and correct arguments

HyukjinKwon commented:
@sun-rui Thanks, I just updated it. There is the [stdout](https://gist.github.com/HyukjinKwon/c34f7b5581de46433a4981f1d30b0421) for this time

SparkQA commented:
**[Test build #59363 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59363/consoleFull)** for PR 13165 at commit [`a1ccd5f`](https://github.com/apache/spark/commit/a1ccd5f383f13a7368516469989527557516ac92).

SparkQA commented:
**[Test build #59363 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59363/consoleFull)** for PR 13165 at commit [`a1ccd5f`](https://github.com/apache/spark/commit/a1ccd5f383f13a7368516469989527557516ac92).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59363/
Test PASSed.

sun-rui commented:
LGTM

sun-rui commented:
could you rebase it to latest master. It is hopeful other test failures will gone

HyukjinKwon commented:
Sure, Thanks!

HyukjinKwon committed:
Fetch upstream

SparkQA commented:
**[Test build #59367 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59367/consoleFull)** for PR 13165 at commit [`7eadefd`](https://github.com/apache/spark/commit/7eadefde6a3d16dbad11de7937042a3c98c2d800).

SparkQA commented:
**[Test build #59367 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59367/consoleFull)** for PR 13165 at commit [`7eadefd`](https://github.com/apache/spark/commit/7eadefde6a3d16dbad11de7937042a3c98c2d800).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59367/
Test PASSed.

HyukjinKwon committed:
Change test context name

SparkQA commented:
**[Test build #59383 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59383/consoleFull)** for PR 13165 at commit [`521f07b`](https://github.com/apache/spark/commit/521f07b704e38675fa0475a0d23c3181a678684b).

SparkQA commented:
**[Test build #59383 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59383/consoleFull)** for PR 13165 at commit [`521f07b`](https://github.com/apache/spark/commit/521f07b704e38675fa0475a0d23c3181a678684b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59383/
Test PASSed.

shivaram commented:
@HyukjinKwon Thanks for the update. I verified this today using a Windows VM and followed the instructions you have written up as well. All the tests passed and the stdout I got is at https://gist.github.com/shivaram/bd3c8bf7466d6cd515f4877f2a9611b3

One minor thing - there is a typo in the last line of https://github.com/apache/spark/blob/branch-2.0/R/WINDOWS.md
`spark.hadoop.fs.defualt.name` should be `spark.hadoop.fs.default.name` (typo in the spelling of `default`). 

Could you make that change as a part of this PR ?

HyukjinKwon commented:
@shivaram Thank you so much! I will keep this machine as it is anyway. Please cc me if there are some PRs dealing with Windows in the future. I can run tests and leave some comments.

HyukjinKwon committed:
Fix a typo

sun-rui commented:
@shivaram

SparkQA commented:
**[Test build #59439 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59439/consoleFull)** for PR 13165 at commit [`752c40a`](https://github.com/apache/spark/commit/752c40a549ce78105db7ea51d06249661fbb5429).

SparkQA commented:
**[Test build #59439 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59439/consoleFull)** for PR 13165 at commit [`752c40a`](https://github.com/apache/spark/commit/752c40a549ce78105db7ea51d06249661fbb5429).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59439/
Test PASSed.

shivaram commented:
Merging this to master and branch-2.0
